### DISCUSSION TITLE
Evaluate creating PK for history table in MYSQL

### DISCUSSION THREAD

**Discussion by bitbp** [TPC-Council/HammerDB#463] [Q&A] [answered]
I can see that other codes (eg. [msssqls](https://github.com/TPC-Council/HammerDB/blob/f89a7298047cd84d398704160c4107b5a67952c1/src/mssqlserver/mssqlsoltp.tcl#L1220) has a PK created but neither MariaDB nor MySQL have it.

Was there any performance validation that led to this decision?
Should it be created to align with the rest of the engines? 

I can see this table is the only one without a PK and MySQL generally speaking benefits from having the PK defined.

**Comment by sm-shaw** [comment, marked as answer]
Hi, Yes MariaDB and MySQL do have a primary key defined, in fact when using InnoDB (the default storage engine) it is essential to have one and if you don't define it you will have one implicitly created.  You can see the primary key defined in the create table statement, for example:
```
set sql(9) "CREATE TABLE `warehouse` (
`w_id` INT(6) NOT NULL,
`w_ytd` DECIMAL(12, 2) NULL,
`w_tax` DECIMAL(4, 4) NULL,
`w_name` VARCHAR(10) BINARY NULL,
`w_street_1` VARCHAR(20) BINARY NULL,
`w_street_2` VARCHAR(20) BINARY NULL,
`w_city` VARCHAR(20) BINARY NULL,
`w_state` CHAR(2) BINARY NULL,
`w_zip` CHAR(9) BINARY NULL,
PRIMARY KEY (`w_id`)
)
```
The key thing to note is that the InnoDB tables are Index-organized tables (or sometimes called clustered tables) where the Index is the table and the data is stored in the index structure. This is probably where the confusion comes from as we are not creating separate indexes here. 
For other storage engines this will not necessarily be the case e.g. the columnstore storage engine has no indexes.

**Comment by bitbp** [comment]
Hello Shaw, thanks for coming back to me. Yes I understand that you need a PK in MySQL and if you don't the engine would have to create an implicit one. And this is exactly why I've created and issue but I think I failed to communicate the problem correctly.

MySQL history table **does not have a PK** 

> CREATE TABLE `history` (
  `h_c_id` INT NULL,
  `h_c_d_id` INT NULL,
  `h_c_w_id` INT NULL,
  `h_d_id` INT NULL,
  `h_w_id` INT NULL,
  `h_date` DATETIME NULL,
  `h_amount` DECIMAL(6, 2) NULL,
  `h_data` VARCHAR(24) BINARY NULL
)




As can be see the code [here](https://github.com/TPC-Council/HammerDB/blob/f89a7298047cd84d398704160c4107b5a67952c1/src/mysql/mysqloltp.tcl) which is why the title of the issue is to evaluate to add that PK definition in the code.

Beside the performance potential gains, if HammerBD would be used to also evaluate performance on systems with replication, a widely known issue is that tables without PK in the replica will underperform.

Could it be possible that the definition for the table history is changed to have an explicit PK? Thanks.

**Reply by sm-shaw** [reply]
Hi, many thanks for the additional explanation that this is only for the history table. With the history table this is what it says in the [TPC-C specification](https://www.tpc.org/tpc_documents_current_versions/pdf/tpc-c_v5.11.0.pdf) that HammerDB TPROC-C is derived from:

_Comment: Rows in the History table do not have a primary key as, within the context of the benchmark, there is no need to uniquely identify a row within this table._

Therefore we do not create a primary key for History in MySQL or MariaDB (or other databases) instead MySQL and MariaDB will auto create an index as follows:

```
+--------------+-----------------+
| table_name   | index_name      |
+--------------+-----------------+
| tpcc/history | GEN_CLUST_INDEX |
+--------------+-----------------+
```

All we do with this table is insert into it as part of the PAYMENT stored procedure, so we do not select from it. I have not observed a performance impact on not having a primary key, however this does not mean that this is not possible as there is some sequence generation required for this default index. 

If you can observe a potential for performance improvement, then you can submit a pull request showing the gains you have measured from the change, and it can then be evaluated.

**Reply by bitbp** [reply]
Hello Steve,

you've mentioned:

> Therefore we do not create a primary key for History in MySQL or MariaDB (or other databases)

But as mentioned on my initial write up I see other engines have it. Like for example [msssqls](https://github.com/TPC-Council/HammerDB/blob/f89a7298047cd84d398704160c4107b5a67952c1/src/mssqlserver/mssqlsoltp.tcl#L1220) that has a PK in the history table defined as `PRIMARY KEY NONCLUSTERED ([h_id])) WITH (MEMORY_OPTIMIZED = ON` and the code on Oracle also has it. 

So or the test either removes PK's from all engines or adds them to all of them but having some engines with it and others without kind of sets a different play ground for the engines and no longer is testing the same.

Regarding performance, as explained on my last post when testing on for example a master replica environment or in a group replication cluster, the lack of primary keys greatly impacts replication performance as the sql thread will have to perform full scans on tables instead of traversing the PK. Here is the [bug report where MySQL ](https://bugs.mysql.com/bug.php?id=53375)clearly explains that the solution it's to force PK's to all the tables.

**Reply by sm-shaw** [reply]
Hi,
The only history table implementation with a primary key is SQL Sever in-memory and this is described in the [HammerDB documentation](https://www.hammerdb.com/docs/ch04s03.html). 
_For a full implementation of in-memory tables a primary key is mandatory, however by definition the HISTORY table does not have a primary key. Therefore, to implement all tables as in-memory an identity column has been added to the HISTORY table._ 
I have double-checked and there is not one I see on Oracle, so having no primary key on HISTORY is correct and is the default across virtually all implementations. 

Having said that, if there is a particular issue that you are experiencing and can quantify the performance impact in a replicated environment on not having a primary key on HISTORY, then you could submit a pull request to fix it.  So just for example, if you shared data that showed adding a primary key to HISTORY improves replication performance by 10% and doesn't impact non-replicated performance and then submitted a PR with your changes, then that would be great. Unfortunately, "greatly impacts replication performance" is not quantified and doesn't give enough reason to make the MySQL/MariaDB implementation different from the other databases at the present time.  Some testing needs to be done to show that this is an actual performance issue. 

Also note that the bug report linked and this particular issue of having no primary key is for the InnoDB storage, however HammerDB does not limit the choice of storage engine, and you could for example use MyRocks and therefore any update to the schema needs to account for all storage engines as well not just InnoDB as a change for one might impact performance on another (and I've seen examples of MyRocks testing with HammerDB and MariaDB very recently).

**Reply by ghost** [reply]
The Value of having a PK on any of these tables is to understand features of Sql Server that depend on a PK - i.e. Transactional Replication. We are trying to use Hammer DB to understand net effect of CDC on Sql Server. One of the use cases we are trying to test is to enable CDC and Transactional Replication. It looks like adding a PK to history table manually would be simple - I am trying to understand the code that updates/inserts into the other tables - have people added PKs to customer and new_order?

**Reply by sm-shaw** [reply]
Hi, yes, look at the in-memory schema for SQL Server as a primary key has been added to history for that, so you can just copy this. Yes, all the other tables already have primary keys.  The use of primary keys has been derived from the TPC-C specification, so follows the guidelines in here [TPC-C specification](https://www.tpc.org/tpc_documents_current_versions/pdf/tpc-c_v5.11.0.pdf).

**Comment by intihar** [comment]
Hi Steve, 
  Just to share one use case. As you probably know, there are some mysql replication solutions that require a PK for replicated tables. This is why we have needed to add the PK to the history table. 
thanks,
